Integrate ABBA to LoHA

monatis opened the pull request:
This PR integrates ABBA to LoHA, based on the discussion [here](https://github.com/huggingface/peft/issues/2587#issuecomment-2995542253).

Raising it to track the progress more easily.

Hyperakan committed:
ENH: Extend LoHa configuration with Khatri-Rao optimization and ABBA feature.

This update introduces additional parameters `r1` and `r2` for specifying separate ranks for the Hadamard components in the LoHa configuration. The `use_khatri_rao` flag is added to enable Khatri-Rao product optimization, reducing memory overhead during weight updates. The initialization method is also enhanced to support ABBA-style initialization, allowing for more efficient weight management. These changes improve flexibility and performance in model training.

monatis commented:
@Hyperakan In addition to type hints and docstrings we talked earlier, could you also test it with the method comparison suite like in this [comment](https://github.com/huggingface/peft/issues/2587#issuecomment-2988544594).
Beside testing with the [current experiment for LoHA](https://github.com/altaidevorg/peft/tree/abba_feature/method_comparison/MetaMathQA/experiments/loha/) we can also add another for ABBA there.

Hyperakan committed:
better docstring and changed string to literal data type for "abba".

BenjaminBossan commented on src/peft/tuners/loha/config.py:
```diff
@@ -69,7 +72,25 @@ class LoHaConfig(LycorisConfig):
+                "If not specified, defaults to r/2 for ABBA-style initialization, or r otherwise."
+            )
+        },
+    )
```
I see that these two parameters are described in the paper, but I think it's too much to have 3 parameters to set the rank. I would suggest either:

1. Drop r1 and r2, always use r1=r2=r.
2. If it's important to have r1!=r2 as an option, I would drop r1, use r and r2 only, where r takes the role of r1 and r2 defaults to r.

Unless I'm missing something, I wouldn't use r/2 as default, as I think it's confusing that changing from LoHa to ABBA halves the effective rank.

BenjaminBossan commented on src/peft/tuners/loha/config.py:
```diff
@@ -86,6 +107,18 @@ class LoHaConfig(LycorisConfig):
         },
     )
+    use_khatri_rao: bool = field(
+        default=False,
```
Note for the future: If we can observe that the numerical difference of using Khatri-Rao is negligible and it is strictly better memory-wise, I wouldn't be opposed to make it the default.

BenjaminBossan commented on src/peft/tuners/loha/layer.py:
```diff
@@ -98,38 +106,158 @@ def reset_adapter_parameters_random(self, adapter_name: str):
+                self.hada_w2_a[adapter_name].data.copy_(U_r2 * fourth_root_S2)
+                self.hada_w2_b[adapter_name].data.copy_(fourth_root_S2.unsqueeze(1) * Vh_r2)
+                
+            except Exception as e:
```
IMO, there is no need to catch the exception. As a user, if I explicitly indicated that I want ABBA and it fails, I would rather get an error than a warning.

BenjaminBossan commented on src/peft/tuners/loha/layer.py:
```diff
@@ -98,38 +106,158 @@ def reset_adapter_parameters_random(self, adapter_name: str):
+        self.scaling[adapter_name] = alpha / r  # Original scaling (for backward compatibility)
+        
+        # ABBA paper: separate scaling factors α/√r₁ and α/√r₂
+        import math
```
Move the import to root level.

BenjaminBossan reviewed (changes requested):
Thanks a lot for working on the ABBA method. This contribution already looks quite clean and well integrated, well done.

I found a couple of smaller issues, please check my comments. Also, as a next step, could you please:

- Add documentation (either extend LoHa or add a separate section for ABBA).
- Extend the tests: For a good start, add some ABBA settings [here](https://github.com/huggingface/peft/blob/87b90f045e84356cbd9bb839f3882b675f362711/tests/test_custom_models.py#L248-L286).

Pinging @RaghavSinghal10 if you have the opportunity, it would be great if you could review, with the focus on the algorithm itself :pray:.

Hyperakan committed:
Refactor parameter handling and update ABBA behavior

- `r` now takes the role of `r1`, with `r2` defaulting to `r`
- Added new parameters to `other_param_names`
- ABBA now raises an error instead of a warning on failure
- Changed `use_khatri_rao` default to "auto":
  - Automatically set to True for ABBA
  - Automatically set to False for LoHa
  - Respects user-defined True/False values

Next steps:
- Add documentation
- Extend test coverage

Hyperakan committed:
-Added abba tests to test_custom_models.py

Hyperakan committed:
Added special tolerance handling for ABBA Conv layers.

BenjaminBossan reviewed (changes requested):
Thanks for the updates, the PR looks quite good. Could you please run `make style` to satisfy ruff?

Also, let's add the documentation for ABBA. One suggestion is that @RaghavSinghal10 writes the docs and I would later add him as co-author, if you're both fine with that suggestion.

> All tests are passing now, and the test coverage can be extended further.

Let's extend the `test_stablediffusion.py` and `test_vision_models.py`. They already have a LoHa config, so just copying it and enabling ABBA should be enough.

Hyperakan commented:
> This works for me - I can go ahead and write the documentation if it works for @Hyperakan

That works for me as well, thanks!

Hyperakan committed:
test: expand coverage in test_stablediffusion.py and test_vision_models.py

Also formatted code using make style.

Hyperakan commented:
Now only the following test is failing:
FAILED tests/test_stablediffusion.py::TestStableDiffusionModel::test_merge_layers_safe_merge[LoHaConfig-config_kwargs2-hf-internal-testing/tiny-sd-pipe] - AssertionError: False, 1.0

I couldn’t identify the root cause of the mismatch. It fails for the regular LoHa configuration (i.e. init_weights = False), but it doesn’t fail in the original codebase.

BenjaminBossan commented on tests/test_stablediffusion.py:
```diff
@@ -288,11 +323,21 @@ def test_merge_layers(self, model_id, config_cls, config_kwargs):
-        assert np.allclose(peft_output, merged_output, atol=1.0)
+        # Increase tolerance for LoHa when init_weights is set to "abba"
+        if config_cls == LoHaConfig and original_config_kwargs.get("text_encoder", {}).get("init_weights") == "abba":
+            atol = 15.0
```
This seems a bit concerning, any idea why ABBA is much more susceptible here?

BenjaminBossan reviewed (changes requested):
Thanks for the updates @Hyperakan. When I run the tests locally with GPU, some ABBA tests with `Conv1D` layers are failing: `pytest tests/test_custom_models.py -k "abba and conv1d"`:

```
FAILED tests/test_custom_models.py::TestPeftCustomModel::test_merge_layers[Conv1d ABBA 3-Conv1dBigger-LoHaConfig-config_kwargs84] - AssertionError
FAILED tests/test_custom_models.py::TestPeftCustomModel::test_safe_merge[Conv1d ABBA 3-Conv1dBigger-LoHaConfig-config_kwargs84] - AssertionError
FAILED tests/test_custom_models.py::TestPeftCustomModel::test_disable_adapters_with_merging[Conv1d ABBA 3-Conv1dBigger-LoHaConfig-config_kwargs84] - AssertionError: assert False
```

This just appears to be a matter of tolerances, so e.g. loosening from 1e-4 to 1e-3 for ABBA should be enough. Could you please check? On CPU, they pass for me.

> Now only the following test is failing: FAILED tests/test_stablediffusion.py::TestStableDiffusionModel::test_merge_layers_safe_merge[LoHaConfig-config_kwargs2-hf-internal-testing/tiny-sd-pipe] - AssertionError: False, 1.0
> 
> I couldn’t identify the root cause of the mismatch. It fails for the regular LoHa configuration (i.e. init_weights = False), but it doesn’t fail in the original codebase.

This was quite a rabbit hole. Indeed, the addition of the ABBA tests results in another LoHa test to fail. That test passes when run in isolation. Therefore, the ABBA test must have some kind of side-effect that makes the LoHa test fail. After some investigation, it's actually caused by this line in `HadaWeightKR.forward`:

`diff_weight = torch.empty(m, n, dtype=w1a.dtype, device=w1a.device)`

For some reason unclear to me, if we replace this with `torch.zeros`, the test passes. At first I thought that the loop could be wrong and some empty data is being returned, which *somehow* stays allocated and thus affects the subsequent tests (LoHa uses `torch.empty` for weight initialization) but AFAICT, there is no issue with the loop. Also, calling `torch.cuda.empty_cache()` between tests also doesn't solve the issue. Therefore, I have no idea what the underlying cause is. Probably it's somewhere else in the LoHa code (or even a PyTorch bug?) but I can't see anything.

For now, I would probably change the `torch.empty` call to `torch.zeros` to be safe, even though that's of course more expensive. LMK if you have any ideas. **Update**: Another idea: collect the chunks in a list and concat it at the end. **Update2**: I tried the idea of collecting and concatenating the chunks, thus avoiding `torch.empty` but the old test still fails, what the heck?? It's got to be some kind of CUDA side effect, as running this on CPU works.

Apart from that, just two small comments.

Hyperakan committed:
fix: remove default values for scale parameters in HadaWeightKR forward method

Additionally, added a fixture in test_stablediffusion.py to disable TF32 in cudnn for consistent testing behavior.

Hyperakan commented:
> @Hyperakan I found a better solution to the failing stable diffusion test:
> 
> ```python
> # test_stablediffusion.py
> import torch
> 
> [...]
> 
> class TestStableDiffusionModel(PeftCommonTester):
>     [...]
> 
>     @pytest.fixture(autouse=True)
>     def autofixture(self):
>         # Disable TF32 in cudnn
>         prev_value = torch.backends.cudnn.allow_tf32
>         try:
>             torch.backends.cudnn.allow_tf32 = False
>             yield
>         finally:
>             torch.backends.cudnn.allow_tf32 = prev_value
>         torch.cuda.empty_cache()
> ```
> 
> For some reason, `allow_tf32=True` causes the issue, by setting it to `False`, all tests pass. No need to change the `torch.empty` call. I also added `torch.cuda.empty_cache()` for good measure but it's not required.

After implementing your suggested solution, I didn’t need to set a special atol value for abba. It's passing all the tests now.

Hyperakan committed:
Merge upstream/main into abba_feature

Resolved conflicts in tests/test_custom_models.py by combining:
- ABBA feature support (init_weights == 'abba')
- OSF feature support (OSFConfig)

Hyperakan commented:
I've resolved the merge conflicts, but I'm still encountering errors in test_custom_models.py , even when running on the upstream/main branch. I'm not sure why this is happening.

Hyperakan committed:
Fix gradient computation in TestPeftCustomModel by adding checks for requires_grad before calling backward and optimizer step.

BenjaminBossan commented:
> I needed to add an if loss.requires_grad check around loss.backward() and optimizer.step() because module_dropout=0.1 in the Vanilla MLP 3 ABBA setup can produce adapter parameters that don't contribute to the computational graph.
> With this change, the tests now pass on CPU.

Ah, nice find, could you please add a comment on each instance on why the check needs to be performed?

@RaghavSinghal10 just a friendly ping if you already had time to work on the docs.

Hyperakan committed:
Added comments to clarify the impact of ABBA initialization on loss gradient requirements.

BenjaminBossan commented:
Gentle ping @RaghavSinghal10

BenjaminBossan commented:
Hey, I wanted to ask if we can revive this PR, as not much was missing. Is anyone of the participants willing to pick this up again?

BenjaminBossan commented:
I revived the idea of adding Khatri-Rao to LoHa in #3560. It's not full on ABBA, but still contains a lot of changes in the same spirit as this PR.

The new PR does not directly build on this PR, but since there was still a lot of ground work laid out here, I would like @monatis and @Hyperakan to be added as co-authors of that PR should be merged. If you two want to be added, please LMK how I should add you. Also, feel free to give feedback on said PR.

monatis commented:
Hi @BenjaminBossan much of the work in this PR belongs to @Hyperakan so you may kindly consider adding him as a co-author. I reviewed #3560 and LGTM. Thank you
